pennylane/devices/default_qubit_jax.py

```diff
@@ -0,0 +1,234 @@
+
+except ImportError as e:  # pragma: no cover
+    raise ImportError("default.qubit.jax device requires installing jax>0.2.0") from e
+
+
+class DefaultQubitJax(DefaultQubit):
+    """Simulator plugin based on ``"default.qubit"``, written using jax.
+
+    **Short name:** ``default.qubit.jax``
+
+    This device provides a pure-state qubit simulator written using jax. As a result, it
+    supports classical backpropagation as a means to compute the gradient. This can be faster than
+    the parameter-shift rule for analytic quantum gradients when the number of parameters to be
+    optimized is large.
+
+    To use this device, you will need to install jax:
+
+    .. code-block:: console
+
+        pip install jax jaxlib
+
+    **Example**
+
+    The ``default.qubit.jax`` is designed to be used with end-to-end classical backpropagation
+    (``diff_method="backprop"``) with the jax interface. This is the default method of
+    differentiation when creating a QNode with this device.
+
+    Using this method, the created QNode is a 'white-box', and is
+    tightly integrated with your jax computation:
+
+    >>> dev = qml.device("default.qubit.jax", wires=1)
+    >>> @qml.qnode(dev, interface="jax", diff_method="backprop")
+    ... def circuit(x):
+    ...     qml.RX(x[1], wires=0)
+    ...     qml.Rot(x[0], x[1], x[2], wires=0)
+    ...     return qml.expval(qml.PauliZ(0))
+    >>> weights = jnp.array([0.2, 0.5, 0.1])
+    >>> # Squeeze is needed since JAX requires scalar outputs for grad.
+    >>> grad_fn = jax.grad(lambda x: circuit(x).squeeze())
+    >>> print(grad_fn(weights))
+    array([-2.2526717e-01 -1.0086454e+00  1.3877788e-17])
+
+    There are a couple of things to keep in mind when using the ``"backprop"``
+    differentiation method for QNodes:
+
+    * You must use the ``"jax"`` interface for classical backpropagation, as jax is
+      used as the device backend.
+
+    .. UsageDetails::
+
+        Jax does randomness is a special way when compared to numpy, in that all PRNGs need to
+        be seeded. While we handle this for you automatically in op-by-op mode, when in a ``jax.jit``
+        function, each call will return the exact same samples if done incorrectly.
+
+        Example:
+
+        .. code-block:: python
+
+            dev = qml.device("default.qubit.jax", wires=1, prng_key=key)
+
+            @jax.jit
+            @qml.qnode(dev, interface="jax", diff_method="backprop")
+            def circuit():
+                qml.Hadamard(0)
+                return qml.sample(qml.PauliZ(wires=0))
+
+            a = circuit()
+            b = circuit() # Wrong! b will be the exact same samples as a.
+
+
+        To fix this, you should wrap your qnode in another function that takes a PRNGKey, and pass
+        that in to your device construction.
+
+        .. code-block:: python
+
+            @jax.jit
+            def keyed_circuit(key):
+                dev = qml.device("default.qubit.jax", interface="jax", prng_key=key)

```

Comment:
Interesting that `@jax.jit` works even with this non-numerical bookkeeping step in the code

Reply:
`jax.jit` will work for any pure function that has jax.DeviceArray inputs and outputs! (It will also still work if there are any side effects in the function, but the side effects are lost after jitting).

Reply:
Nice! Makes sense